### ISSUE TITLE
FlxG.width Reports Unexpected Sizes

### ISSUE THREAD

**Issue by ashes999** [HaxeFlixel/flixel#1432]
**TLDR:** How do I get the screen width for centering an image horizontally and vertically?

I created a simple HaxeFlixel game that tries to center an image on screen. Here are some relevant code snippets:

## Project.xml

```
<!--These window settings apply to all targets-->
	<window width="1024" height="576" fps="60" background="#000000" hardware="true" vsync="true" />
	
<!--Web-specific-->
<window if="web" orientation="portrait"  />
	
<!--Desktop-specific-->
<window if="desktop" orientation="landscape" fullscreen="false" resizable="true" />
	
<!--Mobile-specific-->
<window if="mobile" orientation="landscape" fullscreen="true" width="0" height="0" />
```

## Main.hx

```
var gameWidth:Int = 1024; // Width of the game in pixels (might be less / more in actual pixels depending on your zoom).
var gameHeight:Int = 576; // Height of the game in pixels (might be less / more in actual pixels depending on your zoom).
```

## Play.hx

`trace("FlxG width is " + FlxG.width);`

## Debugging

I'm debugging this in Flash on a 1380x800 screen. Depending on how I resize Chrome, I get different sizes (eg. a long, narrow screen reports a width like 1500).

All this leads me to conclude that **FlxG.width and FlxG.height report the "virtual" size of the game** -- that is, they report the size given the current scale and window size. 

That's probably working as expected, but leaves me with quite a dilemma: if I use this for centering images, they will go off-center when the screen aspect ratio doesn't match the game ratio. How, then, do I center or locate images relative to the expected screen size?

If I use `FlxG.width`, when I have big black bars on the screen, controls get positioned out of the visible space (in the black bars).

If I use a hard-coded constant, which matches what's in `Main.hx`, I get the same problem.

**Comment by Gama11** [comment]
>FlxG.width and FlxG.height report the "virtual" size of the game -- that is, they report the size given the current scale and window size

That's not the case.

What flixel version are you using? What target is this? You mentioned Chrome, so I assume HTML5 or Flash? If HTML5, which backend?

What scalemode? If you didn't change the default, that'd be RatioScaleMode.

**Comment by sruloart** [comment]
This sounds like the normal results of using bitfive with no iframe...

**Comment by Gama11** [comment]
@sruloart This seems to be happening on flash though.

>I'm debugging this in Flash on a 1380x800 screen. Depending on how I resize Chrome [...]

**Comment by sruloart** [comment]
Maybe he's actually using FlxG.game.width/height and not FlxG.width/height...

**Comment by ashes999** [comment]
After several hours of experimenting, I'm still stuck. I will clarify.

- This is on Flash. I don't know about Android, because I don't have sufficient devices to test it, and the emulator just crashes on me.
- I did not change anything else in `Main.hx`

I tried several properties, including, but not limited to:

- `FlxG.width`
- `FlxG.stage.width`
- `openfl.Lib.stage.width`

I am using the default (ratio) scale mode.

Let me also clarify what I mean by "virtual screen size." Imagine your game resolution width is set to 1000x1000px. If you create a chrome window that's 1000x200px, your game is scaled down to 20%; that is, your actual game is displayed as 200x200px.

However, your window is 1000x200px. HaxeFlixel? OpenFL? Lime? Someone? knows that the game is scaled to 20%. Accordingly, the physical coordinate in your browser (1000, 0), at 20% zoom, corresponds to a coordinate of (5000, 0).

---
I don't know how to clearly articulate this. Let me try again.

Whatever's in `Project.xml` and `Main.hx` determine your "virtual" game space. This is the actual, at-100%-zoom pixel size if your game.  The actual, physical size when scaled to my tiny 320x480 phone, is the physical game space.

Knowing the zoom, HaxeFlixel appropriately scales coordinates from physical space to virtual space. That's great.  The problem is, I can't consistently position UI or other elements, neither with hard-coded values of my virtual space, nor physical size of my physical device.

Before this, I expected that HaxeFlixel appropriately scales everything to virtual space, and you get a consistent UI. From testing, I can see that this isn't quite the story.

We can boil this down to one question: I want to position a button 32px from the right edge of the screen (virtual / screen-space), and 32px from the bottom edge of the screen (virtual / screen-space). On my phone, this will map to something like 8px instead of 32.

How do I do this?

**Comment by Tiago-Ling** [comment]
Did you try FlxG.stage.stageWidth and FlxG.stage.stageHeight?

**Comment by ashes999** [comment]
@Tiago-Ling I'm pretty sure I did. I'll try it again if you're convinced it will work.

**Comment by Tiago-Ling** [comment]
I had similar issues in the past with this - sorry for not writing much,
i'm at work right now, but at the time i found that the most reliable way
to get the correct stage size (especially in Flash) is to use
'stage.stageWidth' and 'stage.stageHeight'.

I am not saying that this will solve your problem. You have to take into
consideration the OpenFL window size (in project.xml) and also the flixel
game size (in Main.hx).

For mobile, notice that the project.xml file has a window size of "0,0"
which means the window will scale to whatever the device screen resolution
is.

I suggest you make more tests - i'm not an expert in the matter and i
always use OpenFL window size = Flixel game size for all targets except
mobile in which i use OpenFl window size = 0,0.

What are the values for openfl and flixel you are using?

**Comment by ashes999** [comment]
@Tiago-Ling I'm also at work, so I'll test and get back to you tonight (~10 hours or so).

I don't think I understand the interaction of `Project.xml`, `Main.hx`, and `FlxG`'s width/height (I looked through the source, but it didn't help).

I'll let you know about the versions and I'll retest tonight.

**Comment by ashes999** [comment]
```
flixel-templates: [1.0.2]
flixel-tools: [1.0.4]
flixel: [3.3.6]
hxcpp: [3.1.39]
lime: [2.0.0]
openfl-samples: [2.1.0]
openfl: [2.1.8]
swf: [1.7.1]
```

`FlxG.stage.stageWidth` is better than `FlxG.width` but still gives me an inconsistent width for positioning my UI elements. The range is *much* lower though (1100-1375)

**Comment by ashes999** [comment]
@sruloart how do I consistently position a button 32px from the bottom-right corner of the screen? With the default scale mode, sometimes it appears in the black-bar area, and sometimes on the game content, depending on the device aspect ratio.

The scale mode works great to scale everything down to fit on-screen, but leaves me with no way to consistently position elements.

**Comment by Gama11** [comment]
@ashes999 That black bar is not part of the `FlxGame` / `FlxCamera`, you can't really position anything there. You can:

- change the scale mode so there's no longer a black bar there
- add a separate camera for that button and position it accordingly
- add a flash sprite directly to `FlxG.stage`

**Comment by ashes999** [comment]
@Gama11 when I position a button to `(FlxG.width - button.width - 32, FlxG.height - button.height - 32)`, and the device screen isn't the same aspect ratio as the game size, it shows up on the black area.

**Comment by Gama11** [comment]
@ashes999 Have you tried something like this?

```haxe
package;

import flixel.FlxG;
import flixel.FlxState;
import flixel.ui.FlxButton;
import flixel.util.FlxColor;

class PlayState extends FlxState
{
	var button:FlxButton;
	
	override public function create() 
	{
		bgColor = FlxColor.WHITE;
		add(button = new FlxButton());
	}
	
	override public function onResize(Width:Int, Height:Int):Void 
	{
		var distance = 32 / FlxG.scaleMode.scale.x;
		button.x = FlxG.width - button.width - distance;
		button.y = FlxG.height - button.height - distance;
	}
}
```

**Comment by ashes999** [comment]
Sorry for the delay. 

@Gama11 FlxG.scaleMode.scale doesn't have read access, it seems. I have HaxeFlixel 3.3.6.

**Comment by ashes999** [comment]
I think my request is nonsensical. I have to figure out what I'm trying to achieve.

**Comment by 0xFireball** [comment]
I can confirm that `FlxG.stage.stageWidth` works. I think this is the expected behavior.
